File: reversingLL.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {

        // brute
        // stack<int> st;

        // ListNode* temp = head;

        // while(temp!=NULL){
        //     st.push(temp->val);
        //     temp = temp->next;
        // }

        // temp = head;

        // while(temp!=NULL){
        //     temp->val=st.top();
        //     st.pop();
        //     temp=temp->next;
        // }

        // optimal

        ListNode* temp = head;
        ListNode* prev = NULL;

        while(temp!=NULL){
            ListNode* front = temp->next;

            temp->next = prev;
            prev = temp;
            temp=front;
        }

        return prev;
    }
};

File: reversingLL.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # brute
        temp = head
        st = []

        while(temp):
            st.append(temp.val)
            temp=temp.next

        temp = head

        while(temp):
            temp.val = st.pop()
            temp=temp.next

        return head
        